deviceplugin/usb.go

```diff
@@ -0,0 +1,169 @@
+// Copyright YEAR the generic-device-plugin authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deviceplugin
+
+import (
+	"crypto/sha1"
+	"encoding/binary"
+	"fmt"
+	"k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+	"os"
+	"strconv"
+)
+
+const (
+	usbDevicesDir              = "/sys/bus/usb/devices/"
+	usbDevicesDirVendorIDFile  = "idVendor"
+	usbDevicesDirProductIDFile = "idProduct"
+	usbDevicesDirBusFile       = "busnum"
+	usbDevicesDirBusDevFile    = "devnum"
+	usbDevBus                  = "/dev/bus/usb/%+04d/%+04d"
+)
+
+// UsbSpec represents a USB device specification that should be discovered.
+// A USB device must match exactly on all the given attributes to pass.
+type UsbSpec struct {
+	// Vendor is the USB Vendor ID of the device to match on.
+	// (Both of these get mangled to uint16 for processing - but you should use the hexadecimal representation.)
+	Vendor usbID `json:"vendor"`
+	// Product is the USB Product ID of the device to match on.
+	Product usbID `json:"product"`
+}
+
+type usbID uint16
+
+func (id *usbID) UnmarshalJSON(data []byte) error {
+	if string(data) == "null" || string(data) == `""` {
+		return nil
+	}
+	*id = usbID(binary.LittleEndian.Uint16(data))
+	return nil
+}
+
+func (id *usbID) String() string {
+	return fmt.Sprintf("%04x", int(*id))
+}
+
+// usbDevice represents a physical, tangible USB device.
+type usbDevice struct {
+	// Vendor is the USB Vendor ID of the device.
+	Vendor usbID `json:"vendor"`
+	// Product is the USB Product ID of the device.
+	Product usbID `json:"product"`
+	// Bus is the physical USB bus this device is located at.
+	Bus uint16 `json:"bus"`
+	// BusDevice is the location of the device on the Bus.
+	BusDevice uint16 `json:"busdev"`
+}
+
+func (dev *usbDevice) BusPath() (path string) {
```

Comment:
Since this type isn't used anywhere else, let's keep the methods and fields lowercased
```suggestion
func (dev *usbDevice) busPath() (path string) {
```

Reply:
I'd argue that this is a method that may be used by anything that accesses the usbDevice type (as Bus and BusDevice are already exported), though I acknowledge that usbDevice itself is internal

Reply:
yes this is totally true. I tend to always keep the fields/methods of my unexported types _also_ unexported, though this is not strictly necessary